### ISSUE TITLE
Error while parsing DateTime

### ISSUE THREAD

**Issue by xmascu** [sunsetkookaburra/rust-libesedb#7]
Hi!
I´m using your library to read some ESE Databases for my project and i´m having a issue reading the DateTime values from my ESE database.

The error code that is throwing is: Custom { kind: Other, error: "libesedb_record_get_value_floating_point_64bit: unsupported column type: 8.\nlibesedb_record_get_value_floating_point_64bit: unsupported column type: 8." }

I´m having the problem only with that type of data, the other ones are reading correctly.

If you want more information ask me please.

Thank you so much!!

**Comment by sunsetkookaburra** [comment]
Hello @xmascu, thank for submitting the bug report.

It's good to see you've got some use out of the Rust bindings. I still haven't gotten around to properly making my own Windows FileHistory export tool.

I've taken a look back over my code (of which I realise does not reference things it probably should from the main library) and checked back to the original upstream libyal/libesedb library (specifically `libesedb_record.c`).

I think I have a solution, I'll push that to a branch and get you to try again with the following change in your `Cargo.toml` dependencies and report back, then I can merge the fix and release a new version:
```toml
[dependencies]
libesedb = { git = "https://github.com/sunsetkookaburra/rust-libesedb", rev = "264ac78" }
```

**Comment by xmascu** [comment]
Wow, thank you so much, now it´s working perfectly.
Now is parsing all types of data including the DateTime and not showing any error.

To finish, this is a idea (sorry if i need to post in another issue or in another place) but can you implement the Trait (i think is PartialEq) to compare types of the ESE database.
That will help doing my work a lot parsing the data because i can create cases, for example, a function for DateTime and parsing individually.

Thank you for your fast answer and solution :D

**Comment by sunsetkookaburra** [comment]
No worries, main issue was me assuming how the library would work given the Microsoft documentation (Datetime is a `f64`, but I can't it directly load it as such with the libesedb C library).

**Could you confirm that the date/time values are what you expect?** I don't have a good real-world sample database yet for testing to make sure the decimal-days-since-1900 is correct for the ESE datetime format.

As for the `PartialEq`/`Eq`, that's a good idea. I'll work something up shortly (probably including `TryInto`) and include it along with the datetime fix in the next release which I should have out in the next day or so.

**Comment by xmascu** [comment]
Now i´ve worked with it, is very strange because the datetime given (an f64 data) is weird because it has a lot of zeros:
Return Value: DateTime(9.420157300069825e-300) 
And, if i´m not wrong, usually, the date time is in OLE Automation format and this doesn´t look like an OLE datetime

**Comment by xmascu** [comment]
> Now i´ve worked with it, is very strange because the datetime given (an f64 data) is weird because it has a lot of zeros: Return Value: DateTime(9.420157300069825e-300) And, if i´m not wrong, usually, the date time is in OLE Automation format and this doesn´t look like an OLE datetime

Well, maybe is in FILETIME format but, anyways, too many zeros 😢

**Comment by sunsetkookaburra** [comment]
Interesting, I guess there's some work to do to determine the proper time format. For what it's worth, [the ESE docs](https://learn.microsoft.com/en-us/windows/win32/extensible-storage-engine/jet-coltyp) list `JET_coltypDateTime` (`Value::DateTime`) as:
> A double-precision (8-byte) floating point number that represents a date in fractional days since the year 1900. This column type is identical to the variant date type.

**Comment by sunsetkookaburra** [comment]
I'll try what the [`libyal/libesedb` documentation recommends](https://github.com/libyal/libesedb/blob/main/documentation/Extensible%20Storage%20Engine%20(ESE)%20Database%20File%20(EDB)%20format.asciidoc#user-content-columns), and use `FILETIME`; it seems the Microsoft documentation might be wrong:
> TODO: determine why some documentation refers to JET_coltypDateTime as a double-precision (8-byte) floating point number that represents a date in fractional days since the year 1900. This column type is identical to the variant date type (VT_DATE).

**Comment by sunsetkookaburra** [comment]
@xmascu Could try the latest commit and let me know if things are working now?

**Comment by sunsetkookaburra** [comment]
Above rev assumes Filetime, I'll probably remove assumptions about the format and add options to interpret the value in a couple ways.

**Comment by xmascu** [comment]
I think now we have a better DateTime reading...
Value given: Filetime(133207111534300000)
Translating that value into a DateTime in UTC is like: 2023-02-12T21:32:33.004300Z
I only did a few test but, i think that these values might be correct.

Keep it up my friend and any new changes about the interpretation of values let me know plss

**Comment by sunsetkookaburra** [comment]
That datetime value looks far more reasonable, so I'd say `FILETIME` is correct for at least the majority of cases. I managed to find a couple samples of time data via <https://github.com/libyal/libesedb/blob/main/synctestdata.sh> and <https://github.com/log2timeline/plaso/tree/main/test_data> that follow this.

I'm working on `Into/PartialEq/PartialOrd` traits at the moment for `Value`, and hoping the API can look something like the following:

```rust
let val: Value /* = ... */;

if val < SystemTime::UNIX_EPOCH {
    /* ... */
}
```

and should OleTime (fractional) be needed, explicit conversion:

```rust
if val.to_oletime() < Some(SystemTime::UNIX_EPOCH) {
    /* ... */
}
```

**Comment by sunsetkookaburra** [comment]
@xmascu Hello, I just published `0.1.4` to crates.io so you can set that as your dependency version now.
